### PR TITLE
Add ApplicativeError.redeem and MonadError.redeemWith

### PR DESCRIPTION
**Pull request by joan38**

### CODE REVIEW

File: core/src/main/scala/cats/ApplicativeError.scala [hunk 1]
```diff
@@ -72,6 +72,26 @@ trait ApplicativeError[F[_], E] extends Applicative[F] {
...
+
+  /**
+    * Returns a new value that transforms the result of the source,
+    * given the `recover` or `bind` functions, which get executed depending
+    * on whether the result is successful or if it ends in error.
+    *
+    * This is an optimization on usage of [[attempt]] and [[flatMap]].
+    */
+  def redeemWith[A, B](fa: F[A])(recover: E => F[B], bind: A => F[B]): F[B] =
+    flatMap(attempt(fa))(_.fold(recover, bind)))
```
Comment (joan38) [inline core/src/main/scala/cats/ApplicativeError.scala]: `flatMap` doesn't exist here so what do you think I should do?

**Comment by kailuowang** [comment]
I might've missed something. Is this a shortcut for `fa.map(f).handleError(ef)`?

**Comment by joan38** [comment]
@kailuowang Well it's supposed to be in the type class but because of 2.11 bin incompat it's juts in the ops.
This is to match the `IO.redeem`.

**Comment by kailuowang** [comment]
Thanks, that clarified the intention for me, I think. The issue is that if we can't add it to type class then we can't have instances to override it, then there isn't much point of adding it before 2.0 with which we shall be able to add it to type class without breaking BC.

**Comment by joan38** [comment]
@kailuowang It probably make sense.
How do we handle stuff that will go in 2.0? Do we keep breaking PRs like this open?

**Comment by kailuowang** [comment]
yeah for cases like this , we can add it to the type class, let the mima test fail but schedule it to milesstone 2.0.

**Comment by diesalbla** [comment]
This PR seems to be doing a very similar thing to @alexandru PR here https://github.com/typelevel/cats/pull/2237.

**Comment by joan38** [comment]
@diesalbla Thanks I'm closing this one

**Pull request closed without merging**